refactor: Introduce EvictionManager

dergoegge opened the pull request:
At the moment, the eviction logic is mangled across two different components (`CConnman`, `PeerManager`), so we can't really test it in isolation. This is not completely true for the inbound eviction logic as it exists as static functions in `net.{h.cpp}` for which tests already exist. However, the outbound eviction logic is not covered by any fuzz tests and is only testable by spinning up both a connman and peerman.

This PR splits out the eviction logic into its own component `EvictionManager`. In addition to isolating the eviction logic, we get rid of several layer violations (e.g. `CConnman::ForEachNode/ForNode` calls, `CNode::m_last_block_time`, etc.) between `net` and `net processing`.

One instance of the `EvictionManager` is created at start up and passed as a reference to the connection and peer managers. The connection and peer managers report all eviction relevant information to the eviction manager who ultimately suggests nodes to evict as the result of `EvictionManager::SelectInboundNodeToEvict` and `EvictionManager::SelectOutboundNodesToEvict`.

dergoegge committed:
[net] Remove unused CNode state

Now that the eviction manager is aware of the keyed net group and
whether or not a peer is prefered for eviction, we can get rid of the
corresponding CNode members, since they are not used elsewhere.

dergoegge committed:
[eviction] Create EvictionManager stub

dergoegge committed:
[eviction] Make EvictionManager keep track of candidates

dergoegge committed:
[eviction] Move ping/block/tx times from CNode to EvictionManager

dergoegge committed:
[eviction] Report services to EvictionManager

dergoegge committed:
[eviction] Move BIP37 tx relay and bloom filter status to EvictionManager

dergoegge committed:
[eviction] Move SelectNodeToEvict into EvictionManager

dergoegge committed:
[eviction] Report number of blocks in flight to EvictionManager

dergoegge committed:
[eviction] Report block announce time to EvictionManager

dergoegge committed:
[eviction] Report nodes protected from slow chains to EvictionManager

dergoegge committed:
[net processing] Move outbound eviction logic to EvictionManager

dergoegge committed:
[eviction] Report successful connections to EvictionManager

dergoegge committed:
[eviction] Pass max outbound full/block relay connection counts to EvictionManager

theuni committed:
[refactor] Cleanup EvictionManager::SelectOutboundNodesToEvict

dergoegge committed:
[refactor] Split internals of EvictionManager::SelectOutboundNodesToEvict into two functions

dergoegge committed:
scripted-diff: Rename SelectNodeToEvict for clarification

SelectNodeToEvict only suggests inbound nodes to evict.

-BEGIN VERIFY SCRIPT-
ren() { sed -i "s:\<$1\>:$2:g" $(git grep -l "\<$1\>" ./src ./test); }

ren SelectNodeToEvict SelectInboundNodeToEvict

-END VERIFY SCRIPT-

jnewbery commented:
Concept ACK. Very nice!

jnewbery reviewed (commented):
Assuming this receives enough concept ACKs to move forwards, what do you think about splitting this into 3 PRs:

- move eviction logic into its own translation unit (everything up to 0cc90971d9a1ba57681ece7761746bbe1259421e _[net] Move eviction logic to its own file_)
- introduce `EvictionManager` and use for inbound eviction logic
- use `EvictionManager` for outbound eviction logic.

theuni commented:
Concept ACK!

theuni commented on src/net.h:
```diff
@@ -121,78 +122,6 @@ struct CSerializedNetMsg {
-};
-
-/** Convert ConnectionType enum to a string value */
-std::string ConnectionTypeAsString(ConnectionType conn_type);
```
Why not remove the old implementation here?

theuni commented on src/eviction.cpp:
```diff
@@ -0,0 +1,227 @@
+#include <eviction.h>
```
Missing copyright stuff.

dergoegge replied:
Done.

theuni commented on src/net.cpp:
```diff
@@ -1055,6 +1056,24 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
+    NodeEvictionCandidate candidate{
+        /*id=*/pnode->GetId(),
+        /*m_connected=*/pnode->m_connected,
+        /*m_min_ping_time=*/std::chrono::microseconds::max(),
```
I think we need to justify these values for the sake of review.

For example, a default (unset) value for `m_min_ping_time` could reasonably also be 0, so it's not clear in review why it's set to `max()`. For values that have no real meaning until after version handshake, maybe `std::optional` would better communicate their state?

dergoegge replied:
I'm gonna try to slim down the `NodeEvictionCandidate` interface a bit because a lot of these values don't need to be set by users of the eviction manager. They could just be default initialized internally in the eviction manager. I'm also gonna add some comments explaining the defaults.

I think `std::optional` would make it a bit to complicated but i'll see maybe it works out nicely.

dergoegge replied:
I thought about using `std::optional` here but afaict we would end up treating a `std::nullopt` value for one of these in the same way that we treat them right now w.r.t. to the default values. The default values currently increase the likelihood of eviction for new peers (e.g. we prefer to disconnect peers with slow ping times, so the default is the max possible value) and i have added a comment to document that.

theuni commented on src/net_processing.cpp:
```diff
@@ -2631,6 +2631,7 @@ void PeerManagerImpl::ProcessBlock(CNode& node, const std::shared_ptr<const CBlo
     m_chainman.ProcessNewBlock(block, force_processing, &new_block);
     if (new_block) {
         node.m_last_block_time = GetTime<std::chrono::seconds>();
```
Please save a time and re-use it to avoid the slim chance of setting these to different values.

dergoegge replied:
Done.

theuni commented on src/net_processing.cpp:
```diff
@@ -3450,6 +3450,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             m_orphanage.AddChildrenToWorkSet(tx, peer->m_orphan_work_set);
 
             pfrom.m_last_tx_time = GetTime<std::chrono::seconds>();
+            m_evictionman.UpdateLastTxTime(pfrom.GetId(), GetTime<std::chrono::seconds>());
```
Same thing here, use the same time.

theuni commented on src/eviction.cpp:
```diff
@@ -282,6 +282,14 @@ void EvictionManagerImpl::UpdateLoadedBloomFilter(NodeId id, bool bloom_filter_l
     }
 }
 
+void EvictionManagerImpl::UpdateRelayTxs(NodeId id, bool relay_txs)
```
Since this can never become false again, I think we don't need the bool?

That has the advantage of communicating that it can't be unset.

theuni commented on src/net_processing.cpp:
```diff
@@ -2771,7 +2771,10 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
-            if (fRelay) pfrom.m_relays_txs = true;
+            if (fRelay) {
+                pfrom.m_relays_txs = true;
+                m_evictionman.UpdateRelayTxs(pfrom.GetId(), fRelay);
```
`true` would be more readable, but see above about dropping it entirely.

theuni commented on src/eviction.cpp:
```diff
@@ -302,6 +302,22 @@ void EvictionManagerImpl::UpdateRelayTxs(NodeId id, bool relay_txs)
+{
+    LOCK(m_candidates_mutex);
+    if (const auto& it = m_candidates.find(id); it != m_candidates.end()) {
+        ++it->second.m_blocks_in_flight;
```
This is really confusing to read.

Instead of trying to convince myself that it's correct, I'll ask that you make it more obviously correct:`it->second.m_blocks_in_flight++`

dergoegge replied:
Good point. The dev docs recommend this way but in this case its obviously better to do `foo++`.

theuni commented on src/eviction.cpp:
```diff
@@ -302,6 +302,22 @@ void EvictionManagerImpl::UpdateRelayTxs(NodeId id, bool relay_txs)
+{
+    LOCK(m_candidates_mutex);
+    if (const auto& it = m_candidates.find(id); it != m_candidates.end()) {
+        --it->second.m_blocks_in_flight;
```
Here too.

theuni commented on src/net_processing.cpp:
```diff
@@ -963,6 +963,7 @@ void PeerManagerImpl::RemoveBlockRequest(const uint256& hash)
     state->vBlocksInFlight.erase(list_it);
 
     state->nBlocksInFlight--;
+    m_evictionman.RemoveBlockInFlight(node_id);
```
Since the value is checked in the next line, I assume we're going to want `RemoveBlockInFlight` to return its new value.

dergoegge replied:
`CNodeState::nBlocksInFlight` is not removed in this PR so i think it would be confusing to also do the check with the value from the eviction manager. This is one of those cases that i didn't like because the variables have to be kept in sync. (Arguably it's pretty easy for this one since `nBlocksInFlight` is only modified in two places)

theuni commented on src/net_processing.cpp:
```diff
@@ -993,6 +994,7 @@ bool PeerManagerImpl::BlockRequested(NodeId nodeid, const CBlockIndex& block, st
     std::list<QueuedBlock>::iterator it = state->vBlocksInFlight.insert(state->vBlocksInFlight.end(),
             {&block, std::unique_ptr<PartiallyDownloadedBlock>(pit ? new PartiallyDownloadedBlock(&m_mempool) : nullptr)});
     state->nBlocksInFlight++;
+    m_evictionman.AddBlockInFlight(nodeid);
```
Same for `AddBlockInFlight`

theuni commented on src/test/util/net.cpp:
```diff
@@ -60,6 +60,7 @@ std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(int n_candida
             /*m_network=*/ALL_NETWORKS[random_context.randrange(ALL_NETWORKS.size())],
             /*m_perm_flags=*/NetPermissionFlags::None,
             /*m_conn_type=*/ConnectionType::INBOUND,
+            /*m_blocks_in_flight=*/(int)random_context.randrange(5),
```
I think `MAX_BLOCKS_IN_TRANSIT_PER_PEER` would be more reasonable here?

dergoegge replied:
Yes, i couldn't use the `MAX_BLOCKS_IN_TRANSIT_PER_PEER` constant here but i changed to the right value and left a comment.

theuni commented on src/net_processing.cpp:
```diff
@@ -2277,6 +2278,7 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
 
         if (received_new_header && pindexLast->nChainWork > m_chainman.ActiveChain().Tip()->nChainWork) {
             nodestate->m_last_block_announcement = GetTime();
+            m_evictionman.UpdateLastBlockAnnounceTime(pfrom.GetId(), GetTime<std::chrono::seconds>());
```
Please reuse GetTime.

theuni commented on src/net_processing.cpp:
```diff
@@ -3657,6 +3659,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         // peer's last block announcement time
         if (received_new_header && pindex->nChainWork > m_chainman.ActiveChain().Tip()->nChainWork) {
             nodestate->m_last_block_announcement = GetTime();
+            m_evictionman.UpdateLastBlockAnnounceTime(pfrom.GetId(), GetTime<std::chrono::seconds>());
```
Reuse GetTime here too.

theuni commented on src/net_processing.cpp:
```diff
@@ -2374,6 +2374,7 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, const Peer& peer,
             if (m_outbound_peers_with_protect_from_disconnect < MAX_OUTBOUND_PEERS_TO_PROTECT_FROM_DISCONNECT && nodestate->pindexBestKnownBlock->nChainWork >= m_chainman.ActiveChain().Tip()->nChainWork && !nodestate->m_chain_sync.m_protect) {
                 LogPrint(BCLog::NET, "Protecting outbound peer=%d from eviction\n", pfrom.GetId());
                 nodestate->m_chain_sync.m_protect = true;
+                m_evictionman.UpdateSlowChainProtected(pfrom.GetId(), true);
```
Again, if this cannot become unset, I don't think we need the bool param.

theuni commented on src/net_processing.cpp:
```diff
@@ -2923,6 +2923,7 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::SENDCMPCT, /*high_bandwidth=*/false, /*version=*/CMPCTBLOCKS_VERSION));
         }
         pfrom.fSuccessfullyConnected = true;
+        m_evictionman.UpdateSuccessfullyConnected(pfrom.GetId(), true);
```
Another unneeded bool param.

theuni commented on src/net.cpp:
```diff
@@ -1119,6 +1138,9 @@ void CConnman::DisconnectNodes()
                 m_nodes.erase(remove(m_nodes.begin(), m_nodes.end(), pnode), m_nodes.end());
 
+                // Tell the eviction manager to forget about this node
+                m_evictionman.RemoveCandidate(pnode->GetId());
```
I think this could/should be done outside of `m_nodes_mutex` for the sake of untangling? I'm not actually sure, but if it wouldn't break anything IMO it'd be preferable.

We'd need to create a list of nodes to be evicted, then do the removals outside of the lock.

dergoegge replied:
Done.

(You actually had it like that in your original branch and i changed it because it seemed simpler this way.)

theuni commented on src/net_processing.cpp:
```diff
@@ -4088,7 +4091,9 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
                 tx_relay->m_relay_txs = true;
             }
             pfrom.m_bloom_filter_loaded = true;
+            pfrom.m_relays_txs = true;
```
What's this doing?

theuni commented on src/net_processing.cpp:
```diff
@@ -1269,8 +1269,9 @@ void PeerManagerImpl::AddTxAnnouncement(const CNode& node, const GenTxid& gtxid,
-    CNodeState *state = State(node);
+    CNodeState* state = State(node);
     if (state) state->m_last_block_announcement = time_in_seconds;
+    m_evictionman.UpdateLastBlockAnnounceTime(node, std::chrono::seconds{time_in_seconds});
```
I think this can be done outside of `cs_main` ?

theuni reviewed (changes requested):
Thanks so much for working on this. I took a first pass through all commits except for the last few. I'll need to give those another go with fresh eyes.

Some of these commits are mine, so several of these comments are probably aimed back at myself, but I'm not sparing myself from my review :)

I think we may need to restructure this PR a bit for the sake of readability. Again, I realize that you used my approach to the commit series, but now that I'm trying to review it, it's not so easy to follow.

I think it may be much more reviewable if instead of adding all functionality to EvictionManager and then dropping the old stuff, they were done at the same time. So for ex in a commit that adds ping time to eviction manager, it should be hooked up and used, and the old logic dropped. Imo that would make this series much much easier to understand.

dergoegge commented:
Thanks for the review @theuni! I've addressed most of your comments (still working on https://github.com/bitcoin/bitcoin/pull/25268#discussion_r902986411 and restructuring the commits).

Do you also think it would be good to split this into multiple PRs, like John suggested https://github.com/bitcoin/bitcoin/pull/25268#pullrequestreview-994948655? I think that would also help with readability.

dergoegge committed:
[net] Get candidates from eviction manager

We temporarily introduce a method to get a specific
NodeEvictionCandidate from the eviction manager for use in
`CConnman::AttemptToEvictConnection()`. `EvictionManager::GetCandidate`
is removed once we are done moving the remaining eviction state from
`CNode` to the eviction manager.

jamesob commented:
Is it possible to contextualize the benefit of this relatively large refactor? A lot of code is changing here, and the benefits listed in the PR description (resolution of layer violations) are somewhat abstract.

If this is, for example, more testable, wouldn't it make sense to bundle the added test coverage with these changes so that potential reviewers are able to see the tangible benefit? 

Is there some broader effort that this is part of?

dergoegge commented:
> Is it possible to contextualize the benefit of this relatively large refactor?

I would say this PR is similar to #21148, #19988 or #787 in that it splits some logic (e.g. tx request, orphan handling) into its own component (some of the mentioned PRs also changed behavior which this PR does not), which is good for modularity and testability. At the moment, the eviction logic is mangled across two different components (CConnman, PeerManager), so we can't really test it in isolation. This is not completely true for the inbound eviction logic as it exists as static functions in `net` but the outbound eviction logic is not covered by any fuzz tests and is only testable by spinning up both a connman and peerman.

I also think that isolating this logic will make it easier to reason about potential future changes to it.

The layer violation resolutions are more of a nice side effect and I should have mentioned/stressed the test/fuzz benefits in the PR description.

> If this is, for example, more testable, wouldn't it make sense to bundle the added test coverage with these changes so that potential reviewers are able to see the tangible benefit?

Good idea i will add some fuzz and isolated testing for the outbound eviction logic. 

> Is there some broader effort that this is part of?

Pinging @theuni for this question and to see if he has more to add here in general since this was originally his work.

dergoegge commented:
Rebased and restructured the commits a bit as suggested here: https://github.com/bitcoin/bitcoin/pull/25268#pullrequestreview-1014123554.

theuni commented:
I think @dergoegge explained well, but I'll add a bit on top.

The larger effort, which has been improving over the years, is the notion of self-contained subsystems. As it stands, in order to evict a peer, we need to gather the current states of several otherwise-unrelated things. Not only is this slow and error-prone (real-time locking for each individual state capture), it also makes the eviction logic impossible to reason about (and test) on its own.

So the solution/pattern that's emerged to untangle similar problems in the past is: rather than doing real-time peeks into required subsystems every time we need to do something, instead broadcast the triggering events to a new subsystem that maintains a full view of whatever it needs. In practice, what this means is a producer/consumer pattern where multiple subsystems are notified when _something_ happens, and it's up to each to decide what to do with those changes. Yes, this may mean that several subsystems hold duplicate values for the same thing, but now they don't have to cross over each-other for a full view.

Eventually these patterns could be formalized into something like the signals/slots we have elsewhere, which has the nice side-effect of defining triggers that other subsystems may be interested in as well. But as a first step, imo just untangling the layer violations is a worthy goal.

dergoegge commented:
#25500 has been merged, the PR that splits off the next couple commits is #25572.

jnewbery reviewed (commented):
.

dergoegge committed:
[test] Make all denial of service tests use their own connman/peerman

dergoegge committed:
[doc] Document relevant initial NodeEvictionCandidate values

dergoegge committed:
[fuzz] Add EvictionManager target

dergoegge commented:
Rebased. This now also includes a fuzz target for `EvictionManager`, both inbound and outbound logic is being fuzzed.

dergoegge committed:
[test] Replace DoS eviction tests with direct EvictionManger unit tests
